virttest/libvirt_xml/domcapability_xml.py

```diff
@@ -32,6 +33,106 @@ def __init__(self, virsh_instance=base.virsh):
         self['xml'] = self.__dict_get__('virsh').domcapabilities().stdout.strip()
 
+    def get_additional_feature_list(self, cpumodename):
+        """
+        Get additional CPU features which explicitly specified by <feature>
+        tag in cpu/mode[@name='host-model'] part of virsh domcapabilities.
+
+        In libvirt 3.9 and above, domcapabilities give the overall features supported
+        by current host/qemu/kvm via CPU model and a bunch of additional features. CPU
+        model is one set of features defined by libvirt in '/usr/share/libvirt/cpu_map.xml'
+        Specific features described in domcapabilities are addtional features, the valid
+        policy values are 'require' and 'disable', 'require' means the feature can be emulated
+        by host/qemu/kvm, 'disable' which means feature is disabled by current system due to
+        host CPU doesn't support or qemu/kvm has no ability to emulate it or some other
+        reasons like the feature block migration and so on.
+
+        Below is one snipped CPU host-model part output of domcapabilities.
+        <feature> tag specify addtional feature policy for one certain feature type.
+
+        virsh domcapabilities  | xmllint -xpath "/domainCapabilities/cpu/mode[@name='host-model']" -
+        <mode name="host-model" supported="yes">
+            <model fallback="forbid">Skylake-Client</model>
+            <vendor>Intel</vendor>
+            <feature policy="require" name="ss"/>
+            <feature policy="require" name="hypervisor"/>
+            <feature policy="require" name="tsc_adjust"/>
+            <feature policy="require" name="clflushopt"/>
+            <feature policy="require" name="pdpe1gb"/>
+            <feature policy="require" name="invtsc"/>
+        </mode>
+
+        :param cpumodename: cpu mode name, muste be 'host-model' since libvirt3.9
+        :return: list of features, feature is dict-like, feature name is set to dict key,
+                 feature policy is set to dict value.
+                 returen is like [{'ss': 'require'}, {'pdpe1gb', 'require'}]
+        """
+        feature_list = []  # [{feature1, policy}, {feature2, policy}, ...]
+        xmltreefile = self.__dict_get__('xml')
+        try:
+            for mode_node in xmltreefile.findall('/cpu/mode'):
+                # Get mode which name attribute is 'host-model'
+                if mode_node.get('name') == cpumodename:
+                    for feature in mode_node.findall('feature'):
+                        # Feature invtsc doesn't support migration, so not add it to feature list
+                        item = {}
+                        item[feature.get('name')] = feature.get('policy')
+                        if 'invtsc' not in item:
+                            feature_list.append(item)
+            return feature_list
+        except AttributeError, elem_attr:
+            logging.warn("Failed to find attribute %s" % elem_attr)
+            return []
```

Comment:
you can only return feature_list at last outside the try..except

Reply:
I'll add finally clause to return feature_list. Thanks!